core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -186,8 +186,12 @@ class Partition(val topic: String,
       // record the epoch of the controller that made the leadership decision. This is useful while updating the isr
       // to maintain the decision maker controller's epoch in the zookeeper path
       controllerEpoch = partitionStateInfo.controllerEpoch
-      // add replicas that are new
-      allReplicas.foreach(replica => getOrCreateReplica(replica))
+      // add replicas that are new and reset lastCaughtUpTime of non-isr replicas
+      allReplicas.foreach(replica => {
+        val r = getOrCreateReplica(replica)
+        if (!partitionStateInfo.isr.contains(replica))
+          r.resetLastCaughtUpTime()
+      })
```

Comment:
I think we just need to optimize for the common case. The common case is that a replica is switching from follower to leader and all existing ISRs are expected to continue to be caught up. So, setting lastCaughtUpTime to now for ISRs avoids ISR churns. This is what the original code was trying to do (by initialing lastCaughtUpTimeMsUnderlying to current time when a replica is created).  It is true that in the worse case, a replica will be removed from ISR after 2 x ReplicaLagTimeMaxMs, but that should be rare.